src/profile-logic/import/art-trace.js

```diff
@@ -0,0 +1,932 @@
+// method actions: In sampling-based profiles, those timestamps are the original
+// sample timestamps. A sample that changes the current stack will cause a whole
+// bunch of methods to be exited and entered "simultaneously", i.e. with the
+// same timestamp.
+//
+// This function returns the average of the lowest 20% of timestamp deltas that
+// can be observed among the first 500 method actions.
+function procureSamplingInterval(trace) {
+  const { methodActions } = trace;
+
+  // Gather up to 500 time deltas between method actions on a thread.
+  const deltas: number[] = [];
+  const previousTimestampByThread = new Map();
+  const numberOfActionsToConsider = Math.min(500, methodActions.length);
+  for (let i = 0; i < numberOfActionsToConsider; i++) {
+    const { tid, globalTime } = methodActions[i];
+    const previousTimestamp = previousTimestampByThread.get(tid);
+    if (globalTime !== previousTimestamp) {
+      if (previousTimestamp !== undefined) {
+        const delta = globalTime - previousTimestamp;
+        deltas.push(delta);
+      }
+      previousTimestampByThread.set(tid, globalTime);
+    }
+  }
+
+  // Return the average of the lowest 20% of deltas.
+  deltas.sort((a, b) => a - b);
+  if (deltas.length < 5) {
+    return deltas[0];
+  }
+  const deltasThatMatter = deltas.slice(0, Math.floor(deltas.length / 5));
+  const avg =
+    deltasThatMatter.reduce((prev, cur) => prev + cur, 0) /
+    deltasThatMatter.length;
+  return avg;
+}
+
+class CategoryInfo {
+  idleCategory = 0;
+  otherCategory = 1;
+  blockingCategory = 2;
+  androidCategory = 3;
+  javaCategory = 4;
+  kotlinCategory = 5;
+  androidxCategory = 6;
+  specialCategory = 7;
+  categories = [
+    {
+      name: 'Idle',
+      color: 'transparent',
+    },
+    {
+      name: 'Other',
+      color: 'grey',
+    },
+    {
+      name: 'Blocked',
+      color: 'lightblue',
+    },
+    {
+      name: 'Android',
+      color: 'yellow',
+    },
+    {
+      name: 'Java',
+      color: 'blue',
+    },
+    {
+      name: 'Kotlin / KotlinX',
+      color: 'purple',
+    },
+    {
+      name: 'AndroidX',
+      color: 'orange',
+    },
+    {
+      name: 'Special',
+      color: 'green',
+    },
+  ];
+
+  _specialCategoryInfo;
+
+  constructor(methods) {
+    this._specialCategoryInfo = this.getSpecialCategory(methods);
+    if (this._specialCategoryInfo) {
+      this.categories[
+        this.specialCategory
+      ].name = this._specialCategoryInfo.name;
+    }
+  }
+
+  // Make a category for a frequently-encountered bag of code that is not covered
+  // by the other categories in the category list.
+  // In practice, for profiles obtained from Fenix, we want this to return a
+  // "Mozilla" category for methods on "org.mozilla.*" or "mozilla.*" classes.
+  // This is probably a bit overengineered. But on the plus side, it should work
+  // for non-mozilla code, too.
+  getSpecialCategory(methods) {
```

Comment:
Optional: This seems worth unit testing, especially as others come in to refine these categories in the future.

Reply:
I've added a unit test for this.